File: apps/frontend/lib/workflow/evaluator/evaluate.ts
import {
  isConditionGroup,
  type Condition,
  type ConditionGroup,
  type MissingFieldStrategy,
  type WorkflowItem,
} from '@/lib/types/workflow'
import { applyOperator } from './operators'
import { getFieldValue, type AccessorContext } from './fieldAccessors'

export interface EvaluateOptions {
  missingFieldStrategy?: MissingFieldStrategy
  depth?: number
}

export class MissingFieldError extends Error {
  field: string
  constructor(field: string) {
    super(`Missing field: ${field}`)
    this.field = field
  }
}

export async function evaluate(
  item: WorkflowItem,
  group: ConditionGroup,
  opts: EvaluateOptions = {},
): Promise<boolean> {
  const ctx: AccessorContext = { depth: opts.depth ?? 0 }
  const strategy = opts.missingFieldStrategy ?? 'false'
  return evaluateGroup(item, group, ctx, strategy)
}

async function evaluateGroup(
  item: WorkflowItem,
  group: ConditionGroup,
  ctx: AccessorContext,
  strategy: MissingFieldStrategy,
): Promise<boolean> {
  const initial = group.operator === 'AND'
  let result = initial

  for (const child of group.children) {
    const childResult = isConditionGroup(child)
      ? await evaluateGroup(item, child, ctx, strategy)
      : await evaluateCondition(item, child, ctx, strategy)

    if (group.operator === 'AND') {
      if (!childResult) { result = false; break }
    } else {
      if (childResult) { result = true; break }
      result = false
    }
  }

  return group.negate ? !result : result
}

async function evaluateCondition(
  item: WorkflowItem,
  condition: Condition,
  ctx: AccessorContext,
  strategy: MissingFieldStrategy,
): Promise<boolean> {
  const lookup = getFieldValue(item, condition.field, ctx)
  if (!lookup.found) {
    if (strategy === 'error') throw new MissingFieldError(condition.field)
    if (strategy === 'skip') return false
    return condition.negate ? true : false
  }

  const outcome = applyOperator(
    condition.operator,
    lookup.value,
    condition.value,
    { caseSensitive: condition.options?.caseSensitive ?? true },
  )
  return condition.negate ? !outcome.matched : outcome.matched
}

File: apps/backend/app/modules/workflows/application/conditions/evaluate.py
"""Evaluate a condition group against a WorkflowItem.

Mirrors the frontend evaluator (lib/workflow/evaluator/evaluate.ts) so the in-editor
"Evaluation preview" matches real workflow runs. Conditions are consumed as raw JSON dicts
(the shape persisted by the if-node config), not domain objects.
"""

from typing import Any

from app.modules.workflows.application.conditions.field_accessors import get_field_value, is_found
from app.modules.workflows.application.conditions.operators import apply_operator
from app.modules.workflows.domain.models import WorkflowItem

# Missing-field strategies (see frontend MissingFieldStrategy).
STRATEGY_FALSE = "false"
STRATEGY_ERROR = "error"
STRATEGY_SKIP = "skip"


class MissingFieldError(Exception):
    def __init__(self, field: str) -> None:
        super().__init__(f"Missing field: {field}")
        self.field = field


def evaluate(item: WorkflowItem, group: dict, strategy: str = STRATEGY_FALSE) -> bool:
    return _evaluate_group(item, group, strategy)


def _is_group(node: dict) -> bool:
    return "children" in node


def _evaluate_group(item: WorkflowItem, group: dict, strategy: str) -> bool:
    is_and = group.get("operator") == "AND"
    result = is_and

    for child in group.get("children", []):
        child_result = (
            _evaluate_group(item, child, strategy)
            if _is_group(child)
            else _evaluate_condition(item, child, strategy)
        )
        if is_and:
            if not child_result:
                result = False
                break
        else:
            if child_result:
                result = True
                break
            result = False

    return not result if group.get("negate") else result


def _evaluate_condition(item: WorkflowItem, condition: dict, strategy: str) -> bool:
    negate = bool(condition.get("negate"))
    value = get_field_value(item, condition.get("field", ""))

    if not is_found(value):
        if strategy == STRATEGY_ERROR:
            raise MissingFieldError(condition.get("field", ""))
        if strategy == STRATEGY_SKIP:
            return False
        return True if negate else False

    options = condition.get("options") or {}
    case_sensitive = options.get("caseSensitive", True)
    matched = apply_operator(condition.get("operator", ""), value, condition.get("value"), case_sensitive)
    return not matched if negate else matched


def resolve_strategy(value: Any) -> str:
    return value if value in (STRATEGY_FALSE, STRATEGY_ERROR, STRATEGY_SKIP) else STRATEGY_FALSE
